### ISSUE TITLE
Bitcoin full node via Electrum Server 🌕

### ISSUE THREAD

**Issue by glacierbits** [trezor/trezor-suite#2737]
**Design:** to be added
**Spec:** [Notion](https://www.notion.so/satoshilabs/Full-node-35335fa6d2f543f8853485ed56ed1ded)

---

### Main objective
Add support into [blockchain-link](https://github.com/trezor/trezor-suite/tree/develop/packages/blockchain-link) for Electrum Server. This way the user will have the ability to opt-out from our Blockbook backends and set Electrum server address which they may have running on their personal Bitcoin node. Note that we will support Bitcoin only (no altcoins).

### Blockchain-link
As mentioned the main part of this task is to add a new backend to [blockchain-link](https://github.com/trezor/trezor-suite/tree/develop/packages/blockchain-link). We will either add a new dependency that [@prusnak listed below](https://github.com/trezor/trezor-suite/issues/2737#issuecomment-755145626) or write it ourselves. I have a small preference to write it ourselves because  a) we want to be careful about adding new dependencies b) the projects do not seem very well maintained. I believe the best approach is to get inspired heavily by the libs (if the license allows it) but have it in the blockchain-link directly. But the decision is yours.

### CLI option
We should also not forget about adding a CLI option to launch Suite against the Electrum server directly, otherwise some info might leak to our blockbooks once Suite is started. This needs to be discussed as it might not be needed after #3729?

### Modify Bitcoin settings
User wishing to set their own Electrum server will go to Settings -> Coins -> Bitcoin's Advanced settings where they will be able to replace Blockbook URL with an Electrum server. @matejzak will prepare design for this.

### Useful links
- There are number of public Electrum servers. For example listed [here](https://github.com/spesmilo/electrum/blob/master/electrum/servers.json).
- I have personally tried `electrum.blockstream.info:50002:s`. If you would like to run Electrum against such server you can do that by running `electrum --oneserver --server="electrum.blockstream.info:50002:s"`.

### Next steps
- [ ] Prepare design for settings
- [ ] Discuss support in blockchain-link
- [ ] Add support to blockchain-link
- [ ] Document support in blockchain-link
- [ ] Add CLI option (if agreed upon)
- [ ] Add the option to settings
- [ ] Document the feature as a whole

**Comment by petrhejna** [comment]
Is this an issue covering adding an option to use custom full node (instead of nodes of Satoshi Labs) in the Trezor Suite?

**Comment by prusnak** [comment]
> Is this an issue covering adding an option to use custom full node (instead of nodes of Satoshi Labs)?

You can already use custom node when you run [Blockbook](https://github.com/trezor/blockbook).

This issue is about possibility to connect to Bitcoin Core directly (or at least to Electrum server, which has lower system requirements than Blockbook)

**Comment by josibake** [comment]
Do you have a high level idea on how this will work? specifically, I am running a full node on a raspi with electRS and use electrum desktop as my wallet, connected to my full node. I'd like to use Trezor Suite instead and I'm hoping I can keep the same full node setup and just point Trezor Suite at it once this is implemented.

**Comment by petrhejna** [comment]
I have a same expectations and usecase as @josibake 

I also run Raspi and I am kinda sad. Because these days Suite has some issues with connecring via Tor so having own electrum server would be awesome.


I would even expect Suite not to do ANY other network requests other then towords my node.

**Comment by prusnak** [comment]
We are evaluating this and it indeed seems that connecting to Electrum API is probably the best way forward as this is what most of people have in mind when they say "full node".

Electrum APIs in JS:
* https://www.npmjs.com/package/electrumjs (inactive)
* https://www.npmjs.com/package/electrum-client (inactive)
* https://www.npmjs.com/package/@codewarriorr/electrum-client-js (active)
* https://github.com/BlueWallet/rn-electrum-client (active)
* https://github.com/janoside/electrum-client (active)

**Comment by samuel-emrys** [comment]
@prusnak for what it's worth, I'd be interested in integration with bitcoin core more than electrum.

**Comment by tsusanka** [comment]
@prusnak feel free to check my Tech Spec draft in the original post.

**Comment by prusnak** [comment]
I experimented with https://github.com/janoside/electrum-client and this script works for me:

``` javascript
global.net = require('net');
global.tls = require('tls');

const ElectrumClient = require('electrum-client');

const client = new ElectrumClient(50002, 'electrum.blockstream.info', 'ssl');
console.log('[client]', client);

async function f() {
  const ver = await client.initElectrum({ client: 'trezor-suite', version: '21.10.0' });
  console.log('[ver]', ver);

  // address      = 3QmuBaZrJNCxc5Xs7aGzZUK8RirUT8jRKf
  // scriptPubkey = a914fd37588eedf9475ca000ef6e19627c3a64c6ef1787
  // scripthash   = bytes(reversed(hashlib.sha256(bytes.fromhex(scriptPubkey)).digest())).hex() # <- this is Python
  const scripthash = '495fa456cdb66064db3dae04d7b2f307a874cb6f731ab4251b7d73308001ebba';

  const balance = await client.blockchainScripthash_getBalance(scripthash);
  console.log('[balance]', balance);

  const history = await client.blockchainScripthash_getHistory(scripthash);
  console.log('[history]', history);

  const unspent = await client.blockchainScripthash_listunspent(scripthash);
  console.log('[unspent]', unspent);
}

f();
```

**Comment by hynek-jina** [comment]
- [ ]  Write tests - Electrum backend should returns the same output as the Blockbook

The electrum API returns a maximum of 200 transactions (or addresses)- all in one package. 
- [ ] Check how other applications solve this

**Comment by prusnak** [comment]
> The electrum API returns a maximum of 200 transactions (or addresses)- all in one package.

For the record: This is caused by an electrs setting option. There is an [`index_lookup_limit` variable](https://github.com/romanz/electrs/blob/cd348e72de9071549755170e9709ae922e023083/internal/config_specification.toml#L111-L115) which is set to 200 by default. If this is increased (on the server side), the API will return more than 200 transactions per address.

**Comment by marekrjpolak** [comment]
> For the record: This is caused by an electrs setting option. There is an [`index_lookup_limit` variable](https://github.com/romanz/electrs/blob/cd348e72de9071549755170e9709ae922e023083/internal/config_specification.toml#L111-L115) which is set to 200 by default. If this is increased (on the server side), the API will return more than 200 transactions per address.

Moreover, the problem on `all` seed was caused by too many utxos on some output addresses (e.g. [Bitcoin Eater](https://btc1.trezor.io/address/1BitcoinEaterAddressDontSendf59kuE)) and is currently solved by not calculating the `spent` flag for all tx outputs. Nevertheless the principle is still the same and the issue should be solved by https://github.com/spesmilo/electrumx/pull/90.

**Comment by prusnak** [comment]
I came up with a different solution which is hopefully straightforward and can be done on the server side easily: https://github.com/romanz/electrs/issues/635 - let's introduce a server-side skip-list which would skip known popular addresses.

**Comment by Kixunil** [comment]
Thanks for doing this, folks! If possible it'd be great to be able to set the server *before* any request is made to the default one.

**Comment by prusnak** [comment]
After short discussion, the electrs maintainers decided to disable the `index_lookup_limit` by default: https://github.com/romanz/electrs/pull/637

The change will be released in the upcoming electrs 0.9.4.

This resolves the issue with providing only a maximum of 100/200 transactions per address. We should think about adding some kind of progress in case there is more data being fetched from the electrs server, so an user knows the Suite is not stuck but actually still fetching the data.

**Comment by Kixunil** [comment]
Unfortunately without pagination there's no way to actually measure progress. Maybe show progress by number of addresses. (Will jump around for certain people but I don't believe it can be solved.) This heuristics could improve efficiency:

* Send request for first address (or first few)
* If there are many address reuses scan addresses one-by-one
* Else scan them by chunks

This way you can get a bit more consistent response times, which may look better in the UI. Assuming people either usually reuse addresses or (almost) not at all.

**Comment by prusnak** [comment]
> Unfortunately without pagination there's no way to actually measure progress.

Progress does not have to be a progress bar with percent. Progress can be shown also as a rotating wheel or just a text "please wait, I am doing something" :-)

**Comment by tsusanka** [comment]
We have discussed next steps with @marekrjpolak in more detail:

## Product (@matejzak):
- [x] Create design for Settings - option to add Electrum server (maybe it is the same input as the one for blockbook 🤷)
- [ ] Improve onboarding - option for connecting Electrum server (nice to have, we can roll this out without this as we'll have the CLI `--option`)
- [x] Check and finalise design for "modal loading"
- [x] Create design for signalising that we are using custom backend in the upper menu (similar to how we display that you are using Tor). Note that this might have three states: 1. fully on our blockbook servers 2. fully on custom servers 3. mix of 1 and 2 (using Bitcoin via Electrum server but other coins via our blockbook instances).

## Code (@marekrjpolak):
- [x] Add support for Taproot
- [x] Send info to @matejzak what kind of info we may show in the loader
- [x] Check what we are caching but let's not worry about advanced caching to files (cache only during runtime)
- [x] Discuss Coingecko vs Blockbook:
  - There is a small problem with fetching (historic) fiat rates. We have replaced Blockbook with Electrum server so we do not have the needed data. There are two options:
    1. Hide historic fiat rates and use Coingecko for these users. Note that this has huge implications into the design such as hiding the dashboard, some tool tips etc.
    2. Extrapolate the "fiat rates" logic from blockchain-link into separate module/package/whatever and get fiat rates from Blockbook. This is probably the right solution, although a bit more time consuming.

.. and review obviously.

**Comment by Kixunil** [comment]
FYI regarding Electrum servers Taproot should be no-op because the Electrum protocol already hashes whole `script_verify` and doesn't care what's inside.

**Comment by tsusanka** [comment]
Btw some numbers from testing this on the [PR](https://github.com/trezor/trezor-suite/pull/4500) (commit 5348af3e) on my "small testing seed" (cca 10 transactions):
- About **12 seconds** from choosing Standard wallet to the dashboard loaded up properly on our internal company server. Tor OFF.
- About **22 seconds** from choosing Standard wallet to the dashboard loaded up properly on my personal raspberry pi running Umbrel. Tor OFF.
- About **33 seconds** from choosing Standard wallet to the dashboard loaded up properly on my personal raspberry pi running Umbrel. Tor ON.

**Comment by melroy89** [comment]
This issue is btw very important, since it give me (and most importantly; others as well) more trust in storing bitcoin in your suite. After all, centralized services can become down.

Those services can be temporary down or permanently down (worst case scenario).

**Comment by tsusanka** [comment]
> Discuss Coingecko vs Blockbook:

To recap this: we will keep communicating with Blockbook for fiat rates and we will extend [this blockbook REST endpoint](https://btc1.trezor.io/api/v2/tickers/?timestamp=1638728150) with batching option. @martinboehm will look into it, it should be an easy one :).

**Comment by tsusanka** [comment]
Some product work still needs to be done.

**Comment by bosomt** [comment]
Can you please add update of context help in `Sessings/Crypro` to tasks/next steps ?

<img width="534" alt="obrazek" src="https://user-images.githubusercontent.com/31506317/152787947-56d12b04-3c43-4802-a9d3-35ebe10e0ac7.png">

**Comment by hynek-jina** [comment]
Follow-up issues are tracked in separate [Milestone](https://github.com/trezor/trezor-suite/milestone/22)

**Comment by tsusanka** [comment]
Wooohooo, great job everyone 🎉 🎉 🎉 .
